kube-runtime/src/utils/event_modify.rs

```diff
@@ -0,0 +1,82 @@
+use core::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::task::ready;
+
+use futures::{Stream, TryStream};
+use pin_project::pin_project;
+
+use crate::watcher::{Error, Event};
+
+#[pin_project]
+/// Stream returned by the [`modify`](super::WatchStreamExt::modify) method.
+/// Modifies the [`Event`] item returned by the inner stream by calling
+/// [`modify`](Event::modify()) on it.
+pub struct EventModify<St, F> {
+    #[pin]
+    stream: St,
+    f: F,
+}
+
+impl<St, F, K> EventModify<St, F>
+where
+    St: TryStream<Ok = Event<K>>,
+    F: FnMut(&mut K),
+{
+    pub(super) fn new(stream: St, f: F) -> EventModify<St, F> {
+        Self { stream, f }
+    }
+}
+
+impl<St, F, K> Stream for EventModify<St, F>
+where
+    St: Stream<Item = Result<Event<K>, Error>>,
+    F: FnMut(&mut K),
+{
+    type Item = Result<Event<K>, Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut me = self.project();
+        Poll::Ready(match ready!(me.stream.as_mut().poll_next(cx)) {
+            Some(Ok(event)) => Some(Ok(event.modify(me.f))),
+            Some(Err(err)) => Some(Err(err)),
+            None => None,
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod test {
+    use std::task::Poll;
+
+    use super::{Error, Event, EventModify};
+    use futures::{pin_mut, poll, stream, StreamExt};
+
+    #[tokio::test]
+    async fn eventmodify_modifies_innner_value_of_event() {
+        let restart_val = vec![10];
+        let st = stream::iter([
+            Ok(Event::Applied(0)),
+            Err(Error::TooManyObjects),
+            Ok(Event::Restarted(restart_val)),
+        ]);
+        let ev_modify = EventModify::new(st, |x| {
+            *x += 1;
+        });
+        pin_mut!(ev_modify);
+        assert!(matches!(
+            poll!(ev_modify.next()),
+            Poll::Ready(Some(Ok(Event::Applied(1))))
+        ));
```

Comment:
Nit: No need to use `poll!` if you expect it to be `Ready`.

Reply:
as far as i understand, the `poll!` here is used to get convert the future returned by `ev_modify.next()` into a `Poll` object.

Reply:
Yes, but since it should already be ready it would be equivalent to `ev_modify.next().await`. Then again, on second thought, `poll!` does have the advantage of failing faster if it's not ready for whatever reason.